models/git/branch.go

```diff
@@ -382,20 +393,34 @@ func RenameBranch(ctx context.Context, repo *repo_model.Repository, from, to str
 }
 
 // FindRecentlyPushedNewBranches return at most 2 new branches pushed by the user in 6 hours which has no opened PRs created
-// except the indicate branch
-func FindRecentlyPushedNewBranches(ctx context.Context, repoID, userID int64, excludeBranchName string) (BranchList, error) {
+// doer should not be nil
+// TODO use options to find the branches
+func FindRecentlyPushedNewBranches(ctx context.Context, baseRepo *repo_model.Repository, doer *user_model.User) (BranchList, error) {
 	branches := make(BranchList, 0, 2)
-	subQuery := builder.Select("head_branch").From("pull_request").
+	baseBranch, err := GetBranch(ctx, baseRepo.ID, baseRepo.DefaultBranch)
+	if err != nil {
+		return nil, err
+	}
+	// search all related repos
+	repoCond := builder.Select("id").From("repository").
+		Where(builder.Or(
+			builder.Eq{"id": baseRepo.ID},
+			builder.Eq{"is_fork": true, "fork_id": baseRepo.ID},
+		))
+	// avoid check branches which have already created PRs
+	// TODO add head_branch_id in pull_request table then we can get the branch id from pull_request table directly
+	invalidBranchCond := builder.Select("branch.id").From("branch").
+		InnerJoin("pull_request", "branch.name = pull_request.head_branch AND branch.repo_id = pull_request.head_repo_id").
 		InnerJoin("issue", "issue.id = pull_request.issue_id").
-		Where(builder.Eq{
-			"pull_request.head_repo_id": repoID,
-			"issue.is_closed":           false,
-		})
-	err := db.GetEngine(ctx).
-		Where("pusher_id=? AND is_deleted=?", userID, false).
-		And("name <> ?", excludeBranchName).
+		Where(builder.And(
+			builder.Eq{"issue.is_closed": false},
+			builder.In("pull_request.head_repo_id", repoCond),
+		))
+	err = db.GetEngine(ctx).
+		Where("id != ? AND commit_id != ? AND pusher_id = ? AND is_deleted = ?", baseBranch.ID, baseBranch.CommitID, doer.ID, false).
 		And("updated_unix >= ?", time.Now().Add(-time.Hour*6).Unix()).
```

Comment:
How about moving this condition as the first condition, and repo_id as the second condition

Reply:
Updated the code here: use builder to generate the sql query instead.